### PR TITLE
Optimize performance by replacing a set of nodes with a bit in each node.

### PR DESCRIPTION
**Pull request by davidungar**

### CODE REVIEW

**Comment by davidungar** [comment]
@swift-ci please test

**Comment by davidungar** [comment]
rdar://74466708

**Comment by davidungar** [comment]
@swift-ci please test

**Comment by davidungar** [comment]
@CodaFi I am wondering if the driver still needs this transitive machinery at all. For example: if external dependency E is used by A in a fashion that will change B when E changes, will the frontend record that B depends on E? If so, we could remove all the transitive tracing code.

**Comment by davidungar** [comment]
@swift-ci please test

**Review by nkcsgexi — APPROVED** [review]
LGTM!

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/Node.swift [hunk 1]
```diff
@@ -53,6 +58,14 @@ extension ModuleDependencyGraph {
     }
   }
 }
+
+// MARK: - trace status
+extension ModuleDependencyGraph.Node {
+  var isUntraced: Bool { !isTraced }
+  func beTraced() { isTraced = true }
```
Comment (artemcm) [inline Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/Node.swift:65]: Nit: why not `setTraced`/`setUntraced`? 

Otherwise, LGTM!

**Review by artemcm — APPROVED** [review]

File: Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/Node.swift (hunk 1 above)
Comment (davidungar) [reply Sources/SwiftDriver/IncrementalCompilation/ModuleDependencyGraphParts/Node.swift]: Good idea! Tx for the review. I'll make those changes, retest, and merge.

**Comment by davidungar** [comment]
@swift-ci please test

**Pull request merged**